src/util/triplea/MapXMLCreator/CanalDefinitionsPanel.java

```diff
@@ -0,0 +1,442 @@
+        terrCanals.add(NEW_CANAL_OPTION);
+        currentCanalName = (String) JOptionPane.showInputDialog(null,
+            "Which canal should be selected for territory '" + newTerrName + "?", "Choose Canal",
+            JOptionPane.QUESTION_MESSAGE,
+            null, terrCanals.toArray(new String[terrCanals.size()]), // Array of choices
+            terrCanals.get(0)); // Initial choice
+      }
+      if (terrCanals.isEmpty() || NEW_CANAL_OPTION.equals(currentCanalName)) {
+        String suggestedCanalName;
+        int counter = MapXMLHelper.canalDefinitions.size();
+        do {
+          suggestedCanalName = "Canal" + counter;
+          ++counter;
+        } while (MapXMLHelper.canalDefinitions.keySet().contains(currentCanalName));
+        currentCanalName = JOptionPane.showInputDialog(null,
+            "Which canal should be selected for territory '" + newTerrName + "?", suggestedCanalName);
+
+        while (MapXMLHelper.canalDefinitions.keySet().contains(currentCanalName)) {
+          JOptionPane.showMessageDialog(null, "The canal name " + currentCanalName + " is already in use!",
+              "Input Error", JOptionPane.ERROR_MESSAGE);
+          currentCanalName = JOptionPane.showInputDialog(null,
+              "Which canal should be selected for territory '" + newTerrName + "?", currentCanalName);
+        }
+      } else if (currentCanalName != null) {
+        final Tuple<SortedSet<String>, SortedSet<String>> canalTerrs =
+            MapXMLHelper.canalDefinitions.get(currentCanalName);
+        selectedWaterTerritories = canalTerrs.getFirst();
+        selectedLandTerritories = canalTerrs.getSecond();
+        SwingUtilities.invokeLater(new Runnable() {
+          public void run() {
+            imagePanel.repaint();
+          }
+        });
+        return;
+      }
+      if (currentCanalName == null)
+        return;
+    }
+
+    if (currentCanalName != null) {
+      Tuple<SortedSet<String>, SortedSet<String>> canalTerrs = MapXMLHelper.canalDefinitions.get(currentCanalName);
+      if (canalTerrs == null) {
+        canalTerrs = Tuple.of(new TreeSet<String>(), new TreeSet<String>());
+        MapXMLHelper.putCanalDefinitions(currentCanalName, canalTerrs);
+      }
+
+      selectedWaterTerritories = canalTerrs.getFirst();
+      selectedLandTerritories = canalTerrs.getSecond();
+
+      final SortedSet<String> selectedTerrsSameType;
+      if (newTerrIsWater)
+        selectedTerrsSameType = selectedWaterTerritories;
+      else
+        selectedTerrsSameType = selectedLandTerritories;
+
+      if (selectedTerrsSameType.size() > 0) {
+        final SortedSet<String> commonNeighborsDiffType =
+            getCommonNeighborsOfType(selectedTerrsSameType, !newTerrIsWater);
+        commonNeighborsDiffType.retainAll(newTerrNeighborsDiffType);
+
+        if (commonNeighborsDiffType.size() < 2) {
+          JOptionPane.showMessageDialog(null, "The selected " + (newTerrIsWater ? "water" : "land")

```

Comment:
This message dialog has a complicated message, hard to tell what the params to the method call even. I recommend assigning the message to a variable on the preceeding line, then passing the variable in. Extracting the creation of the message to a method could even be called for (since creating the contents of the message is a different responsibility and also lower level detail than the act of deciding whether to  show a message dialog or not - would be justified in its own method)